## Pandas참조 후 titanic 데이터를 가져오기
- name열이 index로 설정

In [1]:
import pandas as pd

In [6]:
df= pd.read_excel('titanic3.xls', index_col='name')
df

pclass  survived     sex  \
name                                                                        
Allen, Miss. Elisabeth Walton                         1         1  female   
Allison, Master. Hudson Trevor                        1         1    male   
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mr. Hudson Joshua Creighton                  1         0    male   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   
...                                                 ...       ...     ...   
Zabour, Miss. Hileni                                  3         0  female   
Zabour, Miss. Thamine                                 3         0  female   
Zakarian, Mr. Mapriededer                             3         0    male   
Zakarian, Mr. Ortin                                   3         0    male   
Zimmerman, Mr. Leo                                    3         0    male   

                                                     age  sibsp  parch  \
name                                                                     
Allen, Miss. Elisabeth Walton                    29.0000      0      0   
Allison, Master. Hudson Trevor                    0.9167      1      2   
Allison, Miss. Helen Loraine                      2.0000      1      2   
Allison, Mr. Hudson Joshua Creighton             30.0000      1      2   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0000      1      2   
...                                                  ...    ...    ...   
Zabour, Miss. Hileni                             14.5000      1      0   
Zabour, Miss. Thamine                                NaN      1      0   
Zakarian, Mr. Mapriededer                        26.5000      0      0   
Zakarian, Mr. Ortin                              27.0000      0      0   
Zimmerman, Mr. Leo                               29.0000      0      0   

                                                 ticket      fare    cabin  \
name                                                                         
Allen, Miss. Elisabeth Walton                     24160  211.3375       B5   
Allison, Master. Hudson Trevor                   113781  151.5500  C22 C26   
Allison, Miss. Helen Loraine                     113781  151.5500  C22 C26   
Allison, Mr. Hudson Joshua Creighton             113781  151.5500  C22 C26   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  113781  151.5500  C22 C26   
...                                                 ...       ...      ...   
Zabour, Miss. Hileni                               2665   14.4542      NaN   
Zabour, Miss. Thamine                              2665   14.4542      NaN   
Zakarian, Mr. Mapriededer                          2656    7.2250      NaN   
Zakarian, Mr. Ortin                                2670    7.2250      NaN   
Zimmerman, Mr. Leo                               315082    7.8750      NaN   

                                                embarked boat   body  \
name                                                                   
Allen, Miss. Elisabeth Walton                          S    2    NaN   
Allison, Master. Hudson Trevor                         S   11    NaN   
Allison, Miss. Helen Loraine                           S  NaN    NaN   
Allison, Mr. Hudson Joshua Creighton                   S  NaN  135.0   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)        S  NaN    NaN   
...                                                  ...  ...    ...   
Zabour, Miss. Hileni                                   C  NaN  328.0   
Zabour, Miss. Thamine                                  C  NaN    NaN   
Zakarian, Mr. Mapriededer                              C  NaN  304.0   
Zakarian, Mr. Ortin                                    C  NaN    NaN   
Zimmerman, Mr. Leo                                     S  NaN    NaN   

                                                                       home.dest  
name                                             

## 데이터프레임의 columns만 조회

In [16]:
df.columns

Index(['pclass', 'survived', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare',
       'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype='object')

## .drop()을 이용해서 2개의 열을 삭제
- home.dest
- embarked
- sibsp
- 원본에 반영

In [17]:
df.drop(columns=['home.dest','embarked','sibsp'], inplace=True)

## del을 이용한 열 삭제
- parch

In [111]:
del df['parch']

## 사망자 명단 조회하기

In [24]:
df[df['survived']==0]

pclass  survived     sex  \
name                                                                        
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mr. Hudson Joshua Creighton                  1         0    male   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   
Andrews, Mr. Thomas Jr                                1         0    male   
Artagaveytia, Mr. Ramon                               1         0    male   
...                                                 ...       ...     ...   
Zabour, Miss. Hileni                                  3         0  female   
Zabour, Miss. Thamine                                 3         0  female   
Zakarian, Mr. Mapriededer                             3         0    male   
Zakarian, Mr. Ortin                                   3         0    male   
Zimmerman, Mr. Leo                                    3         0    male   

                                                  age  parch    ticket  \
name                                                                     
Allison, Miss. Helen Loraine                      2.0      2    113781   
Allison, Mr. Hudson Joshua Creighton             30.0      2    113781   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0      2    113781   
Andrews, Mr. Thomas Jr                           39.0      0    112050   
Artagaveytia, Mr. Ramon                          71.0      0  PC 17609   
...                                               ...    ...       ...   
Zabour, Miss. Hileni                             14.5      0      2665   
Zabour, Miss. Thamine                             NaN      0      2665   
Zakarian, Mr. Mapriededer                        26.5      0      2656   
Zakarian, Mr. Ortin                              27.0      0      2670   
Zimmerman, Mr. Leo                               29.0      0    315082   

                                                     fare    cabin boat   body  
name                                                                            
Allison, Miss. Helen Loraine                     151.5500  C22 C26  NaN    NaN  
Allison, Mr. Hudson Joshua Creighton             151.5500  C22 C26  NaN  135.0  
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  151.5500  C22 C26  NaN    NaN  
Andrews, Mr. Thomas Jr                             0.0000      A36  NaN    NaN  
Artagaveytia, Mr. Ramon                           49.5042      NaN  NaN   22.0  
...                                                   ...      ...  ...    ...  
Zabour, Miss. Hileni                              14.4542      NaN  NaN  328.0  
Zabour, Miss. Thamine                             14.4542      NaN  NaN    NaN  
Zakarian, Mr. Mapriededer                          7.2250      NaN  NaN  304.0  
Zakarian, Mr. Ortin                                7.2250      NaN  NaN    NaN  
Zimmerman, Mr. Leo                                 7.8750      NaN  NaN    NaN  

[809 rows x 10 columns]

## 시신 식별번호(body)가 없는 데이터 조회하기

In [28]:
no_body_mask = df['body'].isnull()
df[no_body_mask]

pclass  survived     sex  \
name                                                                        
Allen, Miss. Elisabeth Walton                         1         1  female   
Allison, Master. Hudson Trevor                        1         1    male   
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   
Anderson, Mr. Harry                                   1         1    male   
...                                                 ...       ...     ...   
Yousif, Mr. Wazli                                     3         0    male   
Yousseff, Mr. Gerious                                 3         0    male   
Zabour, Miss. Thamine                                 3         0  female   
Zakarian, Mr. Ortin                                   3         0    male   
Zimmerman, Mr. Leo                                    3         0    male   

                                                     age  parch  ticket  \
name                                                                      
Allen, Miss. Elisabeth Walton                    29.0000      0   24160   
Allison, Master. Hudson Trevor                    0.9167      2  113781   
Allison, Miss. Helen Loraine                      2.0000      2  113781   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0000      2  113781   
Anderson, Mr. Harry                              48.0000      0   19952   
...                                                  ...    ...     ...   
Yousif, Mr. Wazli                                    NaN      0    2647   
Yousseff, Mr. Gerious                                NaN      0    2627   
Zabour, Miss. Thamine                                NaN      0    2665   
Zakarian, Mr. Ortin                              27.0000      0    2670   
Zimmerman, Mr. Leo                               29.0000      0  315082   

                                                     fare    cabin boat  body  
name                                                                           
Allen, Miss. Elisabeth Walton                    211.3375       B5    2   NaN  
Allison, Master. Hudson Trevor                   151.5500  C22 C26   11   NaN  
Allison, Miss. Helen Loraine                     151.5500  C22 C26  NaN   NaN  
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  151.5500  C22 C26  NaN   NaN  
Anderson, Mr. Harry                               26.5500      E12    3   NaN  
...                                                   ...      ...  ...   ...  
Yousif, Mr. Wazli                                  7.2250      NaN  NaN   NaN  
Yousseff, Mr. Gerious                             14.4583      NaN  NaN   NaN  
Zabour, Miss. Thamine                             14.4542      NaN  NaN   NaN  
Zakarian, Mr. Ortin                                7.2250      NaN  NaN   NaN  
Zimmerman, Mr. Leo                                 7.8750      NaN  NaN   NaN  

[1188 rows x 10 columns]

살아있는 사람은 시신 식별번호가 없는 것은 당연

## 사망자이나 시신식별번호가 없는 명단 찾기

In [35]:
not_survived_mask = df['survived']==0
df[not_survived_mask&no_body_mask]

pclass  survived     sex  \
name                                                                        
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   
Andrews, Mr. Thomas Jr                                1         0    male   
Baumann, Mr. John D                                   1         0    male   
Baxter, Mr. Quigg Edmond                              1         0    male   
...                                                 ...       ...     ...   
Yousif, Mr. Wazli                                     3         0    male   
Yousseff, Mr. Gerious                                 3         0    male   
Zabour, Miss. Thamine                                 3         0  female   
Zakarian, Mr. Ortin                                   3         0    male   
Zimmerman, Mr. Leo                                    3         0    male   

                                                  age  parch    ticket  \
name                                                                     
Allison, Miss. Helen Loraine                      2.0      2    113781   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0      2    113781   
Andrews, Mr. Thomas Jr                           39.0      0    112050   
Baumann, Mr. John D                               NaN      0  PC 17318   
Baxter, Mr. Quigg Edmond                         24.0      1  PC 17558   
...                                               ...    ...       ...   
Yousif, Mr. Wazli                                 NaN      0      2647   
Yousseff, Mr. Gerious                             NaN      0      2627   
Zabour, Miss. Thamine                             NaN      0      2665   
Zakarian, Mr. Ortin                              27.0      0      2670   
Zimmerman, Mr. Leo                               29.0      0    315082   

                                                     fare    cabin boat  body  
name                                                                           
Allison, Miss. Helen Loraine                     151.5500  C22 C26  NaN   NaN  
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  151.5500  C22 C26  NaN   NaN  
Andrews, Mr. Thomas Jr                             0.0000      A36  NaN   NaN  
Baumann, Mr. John D                               25.9250      NaN  NaN   NaN  
Baxter, Mr. Quigg Edmond                         247.5208  B58 B60  NaN   NaN  
...                                                   ...      ...  ...   ...  
Yousif, Mr. Wazli                                  7.2250      NaN  NaN   NaN  
Yousseff, Mr. Gerious                             14.4583      NaN  NaN   NaN  
Zabour, Miss. Thamine                             14.4542      NaN  NaN   NaN  
Zakarian, Mr. Ortin                                7.2250      NaN  NaN   NaN  
Zimmerman, Mr. Leo                                 7.8750      NaN  NaN   NaN  

[688 rows x 10 columns]

사망자가 809명인데, 식별번호가 없는 명단이 688명이나 된다.

## 사망자이나 시신식별번호가 없는 명단으로 새로운 df 만들기
- `notfound_list`

In [40]:
notfound_list = df[no_body_mask & not_survived_mask]
notfound_list

pclass  survived     sex  \
name                                                                        
Allison, Miss. Helen Loraine                          1         0  female   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1         0  female   
Andrews, Mr. Thomas Jr                                1         0    male   
Baumann, Mr. John D                                   1         0    male   
Baxter, Mr. Quigg Edmond                              1         0    male   
...                                                 ...       ...     ...   
Yousif, Mr. Wazli                                     3         0    male   
Yousseff, Mr. Gerious                                 3         0    male   
Zabour, Miss. Thamine                                 3         0  female   
Zakarian, Mr. Ortin                                   3         0    male   
Zimmerman, Mr. Leo                                    3         0    male   

                                                  age  parch    ticket  \
name                                                                     
Allison, Miss. Helen Loraine                      2.0      2    113781   
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0      2    113781   
Andrews, Mr. Thomas Jr                           39.0      0    112050   
Baumann, Mr. John D                               NaN      0  PC 17318   
Baxter, Mr. Quigg Edmond                         24.0      1  PC 17558   
...                                               ...    ...       ...   
Yousif, Mr. Wazli                                 NaN      0      2647   
Yousseff, Mr. Gerious                             NaN      0      2627   
Zabour, Miss. Thamine                             NaN      0      2665   
Zakarian, Mr. Ortin                              27.0      0      2670   
Zimmerman, Mr. Leo                               29.0      0    315082   

                                                     fare    cabin boat  body  
name                                                                           
Allison, Miss. Helen Loraine                     151.5500  C22 C26  NaN   NaN  
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  151.5500  C22 C26  NaN   NaN  
Andrews, Mr. Thomas Jr                             0.0000      A36  NaN   NaN  
Baumann, Mr. John D                               25.9250      NaN  NaN   NaN  
Baxter, Mr. Quigg Edmond                         247.5208  B58 B60  NaN   NaN  
...                                                   ...      ...  ...   ...  
Yousif, Mr. Wazli                                  7.2250      NaN  NaN   NaN  
Yousseff, Mr. Gerious                             14.4583      NaN  NaN   NaN  
Zabour, Miss. Thamine                             14.4542      NaN  NaN   NaN  
Zakarian, Mr. Ortin                                7.2250      NaN  NaN   NaN  
Zimmerman, Mr. Leo                                 7.8750      NaN  NaN   NaN  

[688 rows x 10 columns]

## notfound_list의 승객 등급별 개수 확인

In [112]:
#모법답안
notfound_list['pclass'].value_counts()

3    473
2    127
1     88
Name: pclass, dtype: int64

In [52]:
#내 답안
pclass1_mask = notfound_list['pclass']==1
pclass2_mask = notfound_list['pclass']==2
pclass3_mask = notfound_list['pclass']==3

88

In [53]:
pclass1_mask.sum()

88

In [54]:
pclass2_mask.sum()

127

In [55]:
pclass3_mask.sum()

473

3등급(하류층)이 많으나, 이건 당연할 수 있다. 비율을 구해보자.

In [65]:
(pclass3_mask.sum()/notfound_list['pclass'].sum() * 100).round(3)

26.86

## 전체 titanic 전체 승객 등급별 개수 확인

In [115]:
#모범답안
df['pclass'].value_counts()


3    709
1    323
2    277
Name: pclass, dtype: int64

In [66]:
#내답안
p1_mask = df['pclass']==1
p2_mask = df['pclass']==2
p3_mask = df['pclass']==3

In [67]:
p1_mask.sum()

323

In [68]:
p2_mask.sum()

277

In [69]:
p3_mask.sum()

709

## 승객 등급별 실종 비율 확인

In [114]:
#모범답안
notfound_list['pclass'].value_counts()/df['pclass'].value_counts()


1    0.272446
2    0.458484
3    0.667137
Name: pclass, dtype: float64

In [73]:
#내답안
(pclass1_mask.sum()/p1_mask.sum()*100).round(2)

27.24

In [74]:
(pclass2_mask.sum()/p2_mask.sum()*100).round(2)

45.85

In [75]:
(pclass3_mask.sum()/p3_mask.sum()*100).round(2)

66.71

## 1. 사망자의 하류층 티겟 가격 분포와 2. 생존자의 하류층 티켓 가격 분포를 확인해보자.

In [109]:
#1. 사망자의 하류층 티겟 가격 분포와
p3_not_survived=df[(df['survived']==0)& (df['pclass']==3)]
p3_not_survived['fare'].describe()

count    527.000000
mean      13.435451
std       12.033986
min        0.000000
25%        7.750000
50%        8.050000
75%       14.500000
max       69.550000
Name: fare, dtype: float64

In [110]:
# 2. 생존자의 하류층 티켓 가격 분포
p3_survived=df[(df['survived']==1)& (df['pclass']==3)]
p3_survived['fare'].describe()

count    181.000000
mean      12.916919
std        9.775015
min        0.000000
25%        7.750000
50%        8.050000
75%       15.741700
max       56.495800
Name: fare, dtype: float64

3등급에서 티켓가격의 차이가 생존유무와 관련성이 있어보이지 않는다.